### DISCUSSION TITLE
Adding drag force for an underwater soft robot arm

### DISCUSSION THREAD

**Discussion by JayLooi** [BJCaasenbrood/SorotokiCode#18] [Q&A] [answered]
Hi @BJCaasenbrood , thanks for making this great work open-sourced. 

I am currently trying to model and simulate an underwater soft robot arm as Cosserat rod using the Shapes class, and would like to include the drag force in the model. However, I notice that the `Shapes.system.fDrag` is always set to 0 inside the `assembleGlobalShapes` method. 

May I know if the drag force is supported in Sorotoki's beam model?

**Comment by BJCaasenbrood** [comment]
Hi @JayLooi, an earlier version of Sorotoki had drag forces included. Since I decided to write a cleaner version of the toolkit, I am slowly implementing older functionality into the newest version.

Thanks for letting me know, if you give me a bit of time, I'd be able to re-implement the function so that you may add drag.

**Reply by BJCaasenbrood** [reply]
Here is an example of an soft beam under drag: https://bjcaasenbrood.github.io/SorotokiCode/model/img/follow.mp4

I'll try my best to implement the functionality by next week.

**Reply by JayLooi** [reply]
Thanks for your prompt response. 

I completely understand that implementing this feature will require time and effort. I'm more than happy to patiently wait for your next update. 

Regarding the earlier version of Sorotoki that you mentioned, could you please let me know if it's available in this repository? I'd appreciate the opportunity to take a look in the meantime.

**Comment by BJCaasenbrood** [comment]
@JayLooi I have a working implementation of viscous drag, I'd expect to push it tomorrow. I'll send you a sample code once it's on the repo.

**Reply by BJCaasenbrood** [reply]
@JayLooi The drag functionality is pushed to the latest version SorotokiModel - https://github.com/BJCaasenbrood/SorotokiModel. If you have the latest version of sorotoki installed, you can call:

```matlab
sorotoki update
```

Otherwise, you can manually overwrite `lib/sorotokimodel`.

**Reply by BJCaasenbrood** [reply]
@JayLooi As for an example, here you go:

```matlab
clr;
M = 25; % number of modes

mat = NeoHookean(0.01,0.4);
mat.params.Zeta = 0.1;
shp = Shapes(pccspace(50,M), [0,M,0,0,0,0], 'Length',175);

shp = shp.setMaterial(mat);
shp = shp.setRadius([8,8,0.85]);

% drag parameters [density_fluid, drag coefficient perpendicular, drag coeff. tangential]
shp = shp.addDrag(900e-12, 0.49, 0.01); % <= New function for adding drag
shp.options.isRampCompensation = true;  % ensures that taper shape is respected in the stiffness/mass
shp.options.Quality = 120;

shp.solver.TimeStep = 1/30;
shp.solver.TimeHorizon = 50;
shp.solver.MaxIteration = 3;

shp.system.Controller = @(x) Control(x);

figure(101); view(30,30);
axis([-120 120 -8 8 -5 150]);
drawnow;

jj = 1;
while shp.solver.Time < shp.solver.TimeHorizon
    shp = shp.update();
    shp = showRenderShapes(shp);

    if jj == 1
        gif('octarm_drag.gif','frame',gcf,'nodither','DelayTime',1/60);
    else
        gif;
    end

    jj = jj + 1;
end 

function tau = Control(shp)
    t = shp.solver.Time;
    tau = zeros(shp.NJoint,1);
    tau(1:5) = -0.75 * sin(3*t);  % add sinusoidal signal to base
end
```

It should produce the following result.

![octarm_drag](https://github.com/BJCaasenbrood/SorotokiCode/assets/36262141/523cbd0b-2239-4991-9c5b-dfb42e47858a)

**Reply by JayLooi** [reply]
Wow, that's fast! Thanks for your quick response on my query, I really appreciate it. 

I have updated my Sorotoki and tried your sample code, it ran as expected. 
Just one small clarification, is the density argument of `addDrag` in $kg\ mm^{-3}$?

**Reply by BJCaasenbrood** [reply]
Yeah, it is! All of Sorotoki uses millimeter instead of meter. So gravitational acceleration is also -9800 mm/s^2. 

ps. I was wondering if you used `sorotoki update` or simply overwrote the Model tool. The update function is barely tested, so knowing if it works would be great! 😃

**Reply by JayLooi** [reply]
Thanks for the clarification. 

Regarding the update, yes I was using `sorotoki update`. And I also ran `sorotoki build --approve` to rebuild the mex files. However, there was an issue with the `computeLagrangianGaussFast_mex`, so I ran `generateMexCLGF` to rebuild it and finally it is working now.

**Reply by BJCaasenbrood** [reply]
Cool happy to hear the update function works! And I'll look into the build function.

Thanks alot!

**Comment by BJCaasenbrood** [comment, marked as answer]
Drag functionality is added with commit: https://github.com/BJCaasenbrood/SorotokiModel/commit/1e385f0c33e55aa34b874f85c588c18eab3b5302.

User can now call `Shapes.addDrag(rho,ct,cp)` to added viscous drag to the beam shapes.
